### DISCUSSION TITLE
Use CMake's FetchContent feature for imports

### DISCUSSION THREAD

**Discussion by lpzantony** [biojppm/rapidyaml#125] [Ideas]
Hello,

First of all, I would like to thank you for writing this library, it solved a lot of performance issues in my projects using YAML.

I am opening this discussion to ask you about the new `FetchContent` feature that arrived with CMake V3.11.

I am slowly transiting to this instead of git sub-modules in my CMake projects and I am wondering if you would be open to use it for this library.

What convinced me to switch to `FetchContent` was [this article](https://medium.com/analytics-vidhya/c-dependency-management-with-cmakes-fetchcontent-4ceca4693a5d) on Medium.

What do you think?

Regards

Antony L

**Comment by biojppm** [comment]
Hey @lpzantony, thanks for reaching out, and apologies for the late reply.

I assume you are suggesting to change the  the current git submodules (eg, c4core) into `FetchContent` directives? If that is your suggestion, I am not inclined to adopt it. 

The submodules I am using are tightly coupled with the repo (ie, they both frequently change), and I want to use git's submodule tracking infrastructure to signify that. It also makes my life much easier when editing the projects. `FetchContent` is very loosely coupled and is less than convenient to edit the dependencies. Also, with submodules the only cost for end-users is having to remember to add `--recursive` when cloning ryml ([a fact which is also referred in the docs](https://github.com/biojppm/rapidyaml#installing-and-using)).

I think this is a very fair trade-off.

(By the way, I do use an approach similar to `FetchContent` when dealing with less-coupled dependencies for development purposes such as googletest or other competing YAML or JSON libraries for benchmark purposes). Again these libraries are not changed by me so in this case you as a user are not even aware of the dependency. Eg, [see the benchmark cmake file](https://github.com/biojppm/rapidyaml/blob/master/bm/CMakeLists.txt).

**Reply by biojppm** [reply]
See also this related issue: #123

**Reply by lpzantony** [reply]
Thank you very much for taking the time to give me such a detailed answer.  This makes a lot of sense. Sometimes , I too wish that `FetchContent` would be more strongly coupled.

Regards
